docs/topics/db/optimization.txt

```diff
@@ -261,44 +261,51 @@ But:
 
 .. _overuse_of_count_and_exists:
 
-Don't overuse ``count()`` and ``exists()``
-------------------------------------------
+Don't overuse ``contains()``, ``count()``, and ``exists()``
+-----------------------------------------------------------
 
 If you are going to need other data from the QuerySet, evaluate it immediately.
 
-For example, assuming an Email model that has a ``subject`` attribute and a
-many-to-many relation to User, the following code is optimal::
+For example, assuming a ``Group`` model that has a many-to-many relation to
+``User``, the following code is optimal::
+
+    if display_group_members:
+        members = group.members.all()
+        if members:
+            if current_user in members:
+                print('You and', len(members) - 1, 'other users are members of this group'.)
+            else:
+                print('There are', len(members), 'members in this group.')
 
-    if display_emails:
-        emails = user.emails.all()
-        if emails:
-            print('You have', len(emails), 'emails:')
-            for email in emails:
-                print(email.subject)
+            for member in members:
+                print(member.username)
         else:
-            print('You do not have any emails.')
+            print('There are no members in this group.')
 
 It is optimal because:
 
 #. Since QuerySets are lazy, this does no database queries if
-   ``display_emails`` is ``False``.
+   ``display_group_members`` is ``False``.
```

Comment:
Whilst the conclusion is correct here, it doesn't follow from the premise. In this example, the lack of database queries has nothing to do with the laziness of Querysets, rather, it is because if `display_group_members` is `False` then no code runs at all!!

Part of me thinks we could illustrate this point better by changing the example to:
```
members = group.members.all()
if display_group_members:
    if members:
        ...
```
The trouble with this suggestion is that the sensible place to put `members = ...` *is* inside the if clause.  The only reason you would put it outside is for the educational purposes of demonstrating a point. Perhaps I am overthinking this...

Reply:
Thanks :+1: I think it's worth moving `members`, this can be considered a part of a larger piece of code.